cpp/src/table/table_rowwise_operators.cuh

```diff
@@ -0,0 +1,319 @@
+
+#include <cudf.h>
+#include "table/device_table.cuh"
+
+namespace
+{
+enum class State
+{
+  False = 0,
+  True = 1,
+  Undecided = 2
+};
+
+struct elements_are_equal
+{
+  template <typename ColumnType>
+  __device__ __forceinline__ bool operator()(gdf_column const &lhs,
+                                             gdf_size_type lhs_index,
+                                             gdf_column const &rhs,
+                                             gdf_size_type rhs_index,
+                                             bool nulls_are_equal = false)
+  {
+    bool const lhs_is_valid{gdf_is_valid(lhs.valid, lhs_index)};
+    bool const rhs_is_valid{gdf_is_valid(rhs.valid, rhs_index)};
+
+    // If both values are non-null, compare them
+    if (lhs_is_valid and rhs_is_valid)
+    {
+      return static_cast<ColumnType const *>(lhs.data)[lhs_index] ==
+             static_cast<ColumnType const *>(rhs.data)[rhs_index];
+    }
+
+    // If both values are null
+    if (not lhs_is_valid and not rhs_is_valid)
+    {
+      return nulls_are_equal;
+    }
+
+    // If only one value is null, they can never be equal
+    return false;
+  }
+};
+} // namespace
+
+/**
+ * @brief  Checks for equality between two rows between two tables.
+ *
+ * @param lhs             The left table
+ * @param lhs_index       The index of the row in the rhs table to compare
+ * @param rhs             The right table
+ * @param rhs_index       The index of the row within rhs table to compare
+ * @param nulls_are_equal Flag indicating whether two null values are considered
+ * equal
+ *
+ * @returns true If the two rows are element-wise equal
+ * @returns false If any element differs between the two rows
+ */
+__device__ inline bool rows_equal(device_table const &lhs,
+                                  const gdf_size_type lhs_index,
+                                  device_table const &rhs,
+                                  const gdf_size_type rhs_index,
+                                  bool nulls_are_equal = false)
+{
+  auto equal_elements = [lhs_index, rhs_index, nulls_are_equal](
+                            gdf_column const &l, gdf_column const &r) {
+    return cudf::type_dispatcher(l.dtype, elements_are_equal{}, l, lhs_index, r,
+                                 rhs_index, nulls_are_equal);
+  };
+
+  return thrust::equal(thrust::seq, lhs.begin(), lhs.end(), rhs.begin(),
+                       equal_elements);
+}
+
+
+/**
+ * @brief  Operator struct that checks for equality between two rows between two tables.
+ */
+struct equality_comparator
+{
+
+/**
+ * @brief  Constructor for equality operator
+ *
+ * @param lhs             The left table
+ * @param nulls_are_equal Flag indicating whether two null values are considered
+ * equal
+ *
+ * Use this constructor when you will be comparing two rows from the same table
+ */
+  equality_comparator(device_table const &lhs, bool nulls_are_equal = false) : _lhs(lhs), _rhs(lhs), _nulls_are_equal(nulls_are_equal)
```

Comment:
Wrap long lines (if I have to scroll right in the GitHub unified diff view, it's too wide :) )

Reply:
done